src/lock/optiql.cpp

```diff
@@ -181,10 +198,30 @@ OptiQL::LockX()  //
 {
   auto qid = GetQID();
   auto *qnode = &(_qnodes[qid]);
+  const auto new_tail = (static_cast<uint64_t>(qid) << kQIDShift) | kXLock;
 
-  throw std::runtime_error{"not implemented yet"};
+  auto cur = lock_.load(kRelaxed);
+  while (true) {
+    qnode->ver.store(cur & kLockMask, kRelaxed);
+    if (lock_.compare_exchange_weak(cur, new_tail, kAcqRel, kRelaxed)) break;
+    CPP_UTILITY_SPINLOCK_HINT
+  }
+
+  auto pred_qid = (cur & kQIDMask) >> kQIDShift;
+  auto *pred_qnode = &(_qnodes[pred_qid]);
+  if ((cur & kLockMask) == kNoLocks) {
+    qnode->ver.store((cur & kVersionMask) + kVersionUpdate, kRelease);
+  } else {
+    // wait until predecessor gives up the lock
+    pred_qnode->next.store(qnode, kRelease);
+    while (qnode->ver.load(kAcquire) & kLockMask) {
+      CPP_UTILITY_SPINLOCK_HINT
+      std::this_thread::yield();
+    }
+    // disable opportunistic read
+    lock_.fetch_and(~(kOPReadLock | kVersionMask), kRelaxed);
```

Comment:
フラグを下ろすだけなら XOR で十分ですね．また，このタイミングで先行スレッドの書いたバージョン値を読む必要があるのと，フェンスを貼るタイミングはこっちですね．

```suggestion
    cur = lock_.fetch_xor(kOPReadLock, kAcquire);
```